docs/designs/termination.md

```diff
@@ -0,0 +1,130 @@
+
+The newly separated reallocation controller will be responsible for electing nodes for termination, which will be explored in future designs.
+
+## Requirements
+* Terminating a node with Karpenter will not leak the instance
+* If termination is requested, the node will eventually terminate
+* Users will be able to implement node deletion safeguards
+* Termination mechanisms will rate-limit at the pod evictions
+
+## Termination Controller
+The termination controller is responsible for gracefully terminating instances. The termination controller draws ideas from the [aws-node-termination-handler](https://github.com/aws/aws-node-termination-handler), [Cluster Autoscaler](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node), and [best practices](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/) on drain given by the Kubernetes community.
+
+### Termination Workflow
+The current termination workflow finds nodes with the label `karpenter.sh/lifecycle-phase: terminable` then cordons, drains, and deletes them. This naively handles a few error cases, has no user safeguards, and makes no effort to rate limit. The new workflow will improve these but continue to monitor nodes with the same labels.
+
+The new termination process will begin with a node that receives a delete request. After Karpenter validates the request, a Karpenter mutating webhook will add the Karpenter [finalizer](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#finalizers) to the node. Then, we cordon and begin draining the node. After no pods remain, we terminate the instance in the cloud provider, then remove the Karpenter finalizer. This will result in the APIServer deleting the node object.
+![](../images/termination-state-machine.png)
+### Triggering Termination
+
+The current termination process acts on a reconcile loop. We will change the termination controller to watch nodes and manage the Karpenter [finalizer](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#finalizers), making it responsible for all node termination and pod eviction logic. 
+
+Finalizers allow controllers to implement asynchronous pre-deletion hooks and are commonly used with CRDs like Karpenter’s Provisioners. Today, a user can call kubectl delete node to delete a node object, but will end up leaking the underlying instance by only deleting the node object in the cluster. We will use finalizers to gracefully terminate underlying instances before Karpenter provisioned nodes are deleted, preventing instance leaking. Relying on kubectl for terminations gives the user more control over their cluster and a Kubernetes-native way of deleting nodes - as opposed to the status quo  of doing it manually in a cloud provider console. 
```

Comment:
we had discussed previously about drawbacks for using finalizers, have those all been ruled out, or can you list some potential drawbacks?

Reply:
We discuss the drawbacks in the final section before the Appendix. We've narrowed down the potential for a drawback here to a very small subset of cases by only creating the finalizer once the request is accepted as valid given the state of the cluster (and potential NDBs). Let me know if you think I should add more on this.

Reply:
Good enough for me :) thanks